contrib_bots/lib/gcalendar.py

```diff
@@ -0,0 +1,262 @@
+from __future__ import print_function
+from datetime import datetime
+import httplib2
+import logging
+import os
+import re
+
+# Uses Google's Client Library
+#   pip install --upgrade google-api-python-client
+from apiclient import discovery, errors
+from oauth2client import client
+from oauth2client import tools
+from oauth2client.file import Storage
+from oauth2client.tools import argparser
+
+# pip install --upgrade dateparser
+import dateparser
+
+SCOPES = 'https://www.googleapis.com/auth/calendar'
+CLIENT_SECRET_FILE = '.gcal_client_secret.json'
+APPLICATION_NAME = 'Zulip Google Calendar Bot'
+
+class UnknownMessageTypeError(Exception):
+    def __init__(self, msgtype):
+        self.type = msgtype
+        self.message = 'Unknown message type "' + msgtype + '"'
+
+    def __str__(self):
+        return self.message
+
+# NOTE ABOUT GOOGLE'S OAUTH
+# Right now there are some duplicates of this function in Zulip.
+#
+# For example, in this case it asks for user's consent to read and write
+# their calendars. Different bots/integrations that require other
+# Google Account permissions, have to repeat this process and create
+# new credentials files, which potentially fills the user's computer with
+# a lot files, spread all over the disk.
+#
+# The proper apporach for this would be to create a script that wraps
+# this procedure, and stores all the credentials in the same place.
+#
+# The ideal situation will be when Zulip becomes capable of managing API
+# keys and user credentials, allowing the user to choose between which
+# bots have access to what. However, this belongs more to the long-term
+# plan in Zulip's roadmap.
+
+def get_credentials():
+    '''Gets valid user credentials from storage.
+
+    If nothing has been stored, or if the stored credentials are invalid,
+    the OAuth2 flow is completed to obtain the new credentials.
```

Comment:
The "stored credentials" are the ones already saved, once you've passed the OAuth2 flow. If you have a wrong `.json`, there's no way the program can know (or at least that I can think of).

If the file `~/.gcalbot-credentials/gcalbot.json` is manually corrupted, the following happens in the log:

```
INFO:googleapiclient.discovery:URL being requested: POST https://www.googleapis.com/calendar/v3/calendars/primary/events?alt=json
INFO:oauth2client.transport:Refreshing due to a 401 (attempt 1/2)
INFO:oauth2client.client:Refreshing access_token
```

After the token is automatically refreshed, it works without any impact for the user.

---

About the revocation token, I've added in my next commit something that should manage this (at least to inform the user of the situation), even though it's rather unlikely having the token revoked while the bot is running.

If it's revoked and then the bot is run, it will just proceed to the OAuth flow, as if it was a "fresh install".